validation-test/stdlib/Range.swift.gyb

```diff
@@ -412,21 +542,180 @@ ${TestSuite}.test("count/whereBoundIsStrideable") {
   expectEqual(0, Bound.timesAdvancedWasCalled.value)
   expectEqual( 0 + (${Self}<Bound>._isHalfOpen ? 0 : 1), range1.count)
   expectEqual(10 + (${Self}<Bound>._isHalfOpen ? 0 : 1), range2.count)
-%   if Self == 'CountableClosedRange':
-  // FIXME: swift-3-indexing-model: implement an O(1) `distance()` on
-  // CountableClosedRange.
-  expectEqual(12, Bound.timesEqualEqualWasCalled.value)
-  expectEqual(2, Bound.timesLessWasCalled.value)
-  expectEqual(0, Bound.timesDistanceWasCalled.value)
-  expectEqual(10, Bound.timesAdvancedWasCalled.value)
-%   else:
   expectEqual(0, Bound.timesEqualEqualWasCalled.value)
   expectEqual(2, Bound.timesLessWasCalled.value)
   expectEqual(2, Bound.timesDistanceWasCalled.value)
   expectEqual(0, Bound.timesAdvancedWasCalled.value)
-%   end
 }
 
+%   if 'Countable' in Self:
+${TestSuite}.test("index(after:)/semantics") {
+  let rangeSizes = [3, 5, 10, 12, 86]
+
+  for endValue in rangeSizes {
+    let start = ${Bound}(0)
+    let end = ${Bound}(endValue)
+    let range: ${Self}<${Bound}> = start${op}end
+    var idx = range.startIndex
+    var previousValue : Int? = nil
+
+    for _ in 0${op}(endValue - 1) {
+      // Advance the index
+      idx = range.index(after: idx)
+      if let previousValue = previousValue {
+        expectEqual(range[idx].value, previousValue + 1)
+      }
+      previousValue = range[idx].value
+    }
+    // Iterate once more to get to the 'after end' position
+    idx = range.index(after: idx)
+    // Now, iterate once more and expect a crash
+    expectCrashLater()
+    _ = range.index(after: idx)
+  }
+}
+
+${TestSuite}.test("index(after:)/semantics/smallestRange") {
+  let range: ${Self}<${Bound}> = ${Bound}(0)${op}${Bound}(0)
+  var idx = range.startIndex
+
+%     if 'Closed' in Self:
+  // Should be able to advance the range once
+  idx = range.index(after: idx)
+%     end
+
+  expectCrashLater()
+    _ = range.index(after: idx)
+}
+
+${TestSuite}.test("index(before:)/semantics") {
+  let rangeSizes = [3, 5, 10, 12, 86]
+
+  for endValue in rangeSizes {
+    let start = ${Bound}(0)
+    let end = ${Bound}(endValue)
+    let range: ${Self}<${Bound}> = start${op}end
+    var idx = range.endIndex
+    var previousValue : Int? = nil
+
+    for _ in 0${op}(endValue - 1) {
+      // Advance the index
+      idx = range.index(before: idx)
+      if let previousValue = previousValue {
+        expectEqual(range[idx].value, previousValue - 1)
+      }
+      previousValue = range[idx].value
+    }
+    // Iterate once more to get to the 'before start' position
```

Comment:
This does not look right. Past-end is a valid index, before-start is not. Valid indices can only be in the following range [startIndex, endIndex).